karapace/coordinator/schema_coordinator.py

```diff
@@ -0,0 +1,935 @@
+
+        # We will only try to perform the rejoin once. If it fails,
+        # we will spin this loop another time, checking for coordinator
+        # and subscription changes.
+        # NOTE: We do re-join in sync. The group rebalance will fail on
+        # subscription change and coordinator failure by itself and
+        # this way we don't need to worry about racing or cancellation
+        # issues that could occur if re-join were to be a task.
+        success = await self._do_rejoin_group()
+        if success:
+            self._start_heartbeat_task()
+            return True
+        return False
+
+    def _start_heartbeat_task(self) -> None:
+        if self._heartbeat_task is None:
+            self._heartbeat_task = create_task(self._heartbeat_routine())
+
+    async def _stop_heartbeat_task(self) -> None:
+        if self._heartbeat_task is not None:
+            if not self._heartbeat_task.done():
+                self._heartbeat_task.cancel()
+                await self._heartbeat_task
+            self._heartbeat_task = None
+
+    async def _heartbeat_routine(self) -> None:
+        last_ok_heartbeat = time.monotonic()
+        hb_interval = self._heartbeat_interval_ms / 1000
+        session_timeout = self._session_timeout_ms / 1000
+        retry_backoff = self._retry_backoff_ms / 1000
+        sleep_time = hb_interval
+
+        # There is no point to heartbeat after Broker stopped recognizing
+        # this consumer, so we stop after resetting generation.
+        while self.member_id != JoinGroupRequest[0].UNKNOWN_MEMBER_ID:
+            try:
+                await asyncio.sleep(sleep_time)
+                await self.ensure_coordinator_known()
+
+                t0 = time.monotonic()
+                success = await self._do_heartbeat()
+            except asyncio.CancelledError:
+                break
+
+            # NOTE: We let all other errors propagate up to coordination
+            # routine
+
+            if success:
+                last_ok_heartbeat = time.monotonic()
+                sleep_time = max((0, hb_interval - last_ok_heartbeat + t0))
+            else:
+                sleep_time = retry_backoff
+
+            session_time = time.monotonic() - last_ok_heartbeat
+            if session_time > session_timeout:
+                # the session timeout has expired without seeing a successful
+                # heartbeat, so we should probably make sure the coordinator
+                # is still healthy.
+                LOG.error("Heartbeat session expired - marking coordinator dead")
+                self.coordinator_dead()
+
+        LOG.debug("Stopping heartbeat task")
+
+    async def _do_heartbeat(self) -> bool:
+        version = 0 if self._client.api_version < (0, 11, 0) else 1
+        request = HeartbeatRequest[version](self.group_id, self.generation, self.member_id)
+        LOG.debug("Heartbeat: %s[%s] %s", self.group_id, self.generation, self.member_id)
```

Comment:
ok seems to be implicit, so I think we should be sure that the server understood that we want to cancel a subscription